common.go

```diff
@@ -1,71 +1,171 @@
 package slinguist
 
+import (
+	"path/filepath"
+	"strings"
+)
+
+// OtherLanguage is used as a zero value when a function can not return a specific language.
 const OtherLanguage = "Other"
 
-var (
-	ExtensionsByLanguage map[string][]string
-	ignoredExtensions    = map[string]bool{
-		".asc": true, ".cgi": true, ".fcgi": true, ".gml": true, ".fx": true,
-		".vhost": true,
+var strategies = []func(filename string, content []byte) (languages []string){
+	getLanguagesByModeline,
+	getLanguagesByFilename,
+	getLanguagesByShebang,
+	getLanguagesByExtension,
+	getLanguagesByContent,
+}
+
+// GetLanguage applies a sequence of strategies based on the given filename and content
+// to find out the most probably language to return.
+func GetLanguage(filename string, content []byte) string {
+	candidates := make([]string, 0, 5)
+	for _, strategy := range strategies {
+		languages := strategy(filename, content)
+		if len(languages) == 1 {
+			return languages[0]
+		}
+
+		if len(languages) > 0 {
+			candidates = append(candidates, languages...)
```

Comment:
You are right, it doesn't affect to overcome but process twice the same candidate. Maybe the Classify interface must takes in a map[language]weight, with weight as an int that stores the number of times the candidate has been seen. The weight could be used in an alternative classifier that wants consider a pre-conditions dynamically (like the number of times the candidates appeared in the chain of strategies). What do you think @smola ?